iceoryx_posh/test/moduletests/test_popo_publisher_port.cpp

```diff
@@ -0,0 +1,390 @@
+// Copyright (c) 2020 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iceoryx_posh/iceoryx_posh_types.hpp"
+#include "iceoryx_posh/internal/mepoo/memory_manager.hpp"
+#include "iceoryx_posh/internal/popo/building_blocks/chunk_queue_data.hpp"
+#include "iceoryx_posh/internal/popo/building_blocks/chunk_queue_popper.hpp"
+#include "iceoryx_posh/internal/popo/ports/publisher_port_data.hpp"
+#include "iceoryx_posh/internal/popo/ports/publisher_port_roudi.hpp"
+#include "iceoryx_posh/internal/popo/ports/publisher_port_user.hpp"
+
+#include "iceoryx_posh/mepoo/mepoo_config.hpp"
+#include "iceoryx_utils/error_handling/error_handling.hpp"
+#include "iceoryx_utils/internal/posix_wrapper/shared_memory_object/allocator.hpp"
+#include "test.hpp"
+
+#include <memory>
+
+using namespace ::testing;
+
+struct DummySample
+{
+    uint64_t dummy{42};
+};
+
+class PublisherPort_test : public Test
+{
+  protected:
+    PublisherPort_test()
+    {
+        m_mempoolconf.addMemPool({CHUNK_SIZE, NUM_CHUNKS_IN_POOL});
+        m_memoryManager.configureMemoryManager(m_mempoolconf, &m_memoryAllocator, &m_memoryAllocator);
+    }
+
+    ~PublisherPort_test()
+    {
+    }
+
+    void SetUp()
+    {
+    }
+
+    void TearDown()
+    {
+    }
+
+    static constexpr size_t MEMORY_SIZE = 1024 * 1024;
+    uint8_t m_memory[MEMORY_SIZE];
+    static constexpr uint32_t NUM_CHUNKS_IN_POOL = 20;
+    static constexpr uint32_t CHUNK_SIZE = 128;
+
+    iox::posix::Allocator m_memoryAllocator{m_memory, MEMORY_SIZE};
+    iox::mepoo::MePooConfig m_mempoolconf;
+    iox::mepoo::MemoryManager m_memoryManager;
+
+    // publisher port w/o history
+    iox::popo::PublisherPortData m_publisherPortData{
+        iox::capro::ServiceDescription("a", "b", "c"), "myApp", &m_memoryManager};
+    iox::popo::PublisherPortRouDi m_publisherPortRouDi{&m_publisherPortData};
+    iox::popo::PublisherPortUser m_publisherPortUser{&m_publisherPortData};
+};
+
+TEST_F(PublisherPort_test, initial_state)
+{
+    EXPECT_FALSE(m_publisherPortUser.isOffered());
+    EXPECT_FALSE(m_publisherPortUser.hasSubscribers());
+    auto maybeCaproMessage = m_publisherPortRouDi.getCaProMessage();
+    EXPECT_FALSE(maybeCaproMessage.has_value());
+}
+
+TEST_F(PublisherPort_test, offering_stop_offering)
+{
+    // initial state
+    EXPECT_FALSE(m_publisherPortUser.isOffered());
```

Comment:
That the initial state has no offered service and no capro message received was already tested in the first unit test and is working therefore! Please do not test it again since in this test you would like the verify that offering is working.